cmd/containerboot/main.go

```diff
@@ -306,81 +313,102 @@ authLoop:
 			log.Fatalf("error creating new netfilter runner: %v", err)
 		}
 	}
+	notifyChan := make(chan ipn.Notify)
+	errChan := make(chan error)
+	go func() {
+		for {
+			n, err := w.Next()
+			if err != nil {
+				errChan <- err
+			} else {
+				notifyChan <- n
+			}
+		}
+	}()
 	for {
-		n, err := w.Next()
-		if err != nil {
+		select {
+		case <-ctx.Done():
+			// Although killTailscaled() is deferred earlier, if we
+			// have started the reaper defined below, we need to
+			// kill tailscaled and let reaper clean up child
+			// processes.
+			killTailscaled()
+			return
```

Comment:
The comment above says that we need to let the reaper clean up child processes, but I don't think we have any synchronization in place that would actually guarantee this. As I understand, `return` here will race with the reaper, and there's a chance that the `containerboot` process will terminate before it gets the SIGCHLD about `tailscaled`.

I suspect we don't actually need to reap tailscaled after it's done, since we are exiting anyway, and the whole container will terminate imminently. So perhaps simply `return` here will be sufficient, letting `killTailscaled` run as part of defer?

Reply:
I read around a bit and my understanding is that the container orchestrator won't necessarily reap zombies

https://github.com/krallin/tini/issues/8#issuecomment-146135930 (couldn't find a more authoritative source)

Reply:
> the container orchestrator won't necessarily reap zombies

I believe if `containerboot` runs as PID 1 in a given PID namespace, and exits, then the kernel will destroy the PID namespace alongside any unreaped zombies. If we share the PID namespace with other processes, then it will be `init`'s job to reap it, so it's good to try to do that ourselves. Adding a wg to wait for the reaper makes sense, thanks!